### PR TITLE
Split step

### PR DESCRIPTION
**Pull request by bgobbi**
Hi Lee-Ping,
to get this working for the NNP work I had to do split what we had in optimizer.step() into three parts:
- ```step(self, optObj)``` now does only everything up to taking the optimization step, it returns the modified coordinates
- ```optObj.calcEnergyForce()``` is unchanged but needs to be called after ```step()``` and before ```evaluateStep()```
- ```evaluateStep(self, optObj)``` evaluates the last step taken and returns the status

The [Optimize](https://github.com/bgobbi/geomeTRIC/blob/e39e704a7b99886b8753f54149709de62c645645/geometric/optimize.py#L1363) method now looks like this:
```
    optzer = Optimizer(params, xyzout, xyzout2);
    optObj  = OptObject(coords, molecule, IC, engine, params.trust, dirname)
    
    # take a step based on grad in optObj
    optzer.step(optObj)
    
    ### Calculate new Energy and Gradient ###
    optObj.calcEnergyForce()
    
    while optzer.evaluateStep() is OPT_RESULT.NOT_CONVERGED:
        optzer.step(optObj)
        optObj.calcEnergyForce()
        
    return optObj.progress
```

I had not realized that ```step()``` was calling ```calcEnergyForce()```. To be able to calculate the energies and forces in batch I need to be able to run this independently from the step and evaluation method.
Would you mind taking a look and seeing if this makes sense to you?

I ran the water2 example and got identical results to the current master branch. I consider this changes relative low-medium risk as all I have done is spliting the method and moving the following variables from the ```step()``` method to the ```optObject``` class:
```
farConstraints
conSatisfied  
cnorm
Eprev
expect
dy   
Yprev
Xprev
Gprev
```
I would also like to make another suggestion (but this is for a later pull request). Would you be OK with replacing print statements that currently go to stdout with print statements that go to stderr? All our programs work by piping and for that to work only output data may got to stdout all log messages need to go to stderr.

Thanks for taking a look.
Alberto

### CODE REVIEW

**Comment by leeping** [comment]
Hello Alberto,

Thanks for the changes.  I think overall it looks clear and logical, and I have two main questions:

1) I'm not sure which variables belong in `Optimizer` vs. in `OptObject`.  Is there a guiding philosophy for which class a given variable should live in (e.g. `Gprev`), or are they being moved between the two classes mainly out of necessity?  I am not sure whether this could make variables harder to find in the future.

2) A more general question is whether future applications might necessitate further mid- to large-scale reorganizations of the code in order to separate out a few methods that need to be called in a different way.  Is it possible to accomplish what you need with some combination of hooks and asynchronous execution?  For example, your `Optobj.Engine.calc()` implementation might add a single structure into the batch evaluation for the NNP and then wait; the batch executes when it gets the structures from all of the `Optobj` instances; then all of the calc() objects retrieve their energies and forces from the batch and return.

Thanks,

- Lee-Ping

**Comment by bgobbi** [comment]
1) My design was that ```Optimizer``` should contain only the algorithm not the data. In the case of a batch minimization my goal was to have a single ```Optimizer``` and one ```ObtObject``` per conformation. So essentially anything that can be different between two confromers/molecules being minimized would need to be in ```OptObject```.

3) I was thinking about that, but am not sure how to make this work unless I would create one thread per ```OptObject```. The thread would then stop and wait in ```Optobj.Engine.calc()``` until the energies and forces are present, then continue. The NNP thread could wait for n requests and then compute the energy and forces. However, having one thread per ```OptObject``` seems too expensive.
Do you see a different way?

**Comment by bgobbi** [comment]
The current code has still a Bug:
https://github.com/bgobbi/geomeTRIC/blob/e39e704a7b99886b8753f54149709de62c645645/geometric/optimize.py#L1127
and
https://github.com/bgobbi/geomeTRIC/blob/e39e704a7b99886b8753f54149709de62c645645/geometric/optimize.py#L1147

return different objects. I am still working on that. Probably returning the status in 1147 and making the coordinates accessible via ```OptObject```.

**Comment by leeping** [comment]
1.  I agree it's a good design principle that the Optimizer should not contain any information that is potentially different between two OptObjects.  Also, I think the reorganization of the optimization step into (1) taking the optimization step, (2) computing the energy / gradient, and (3) checking the convergence criteria is basically sound.  

I intend to merge your contributions in this PR after a few days, after we've had some discussions and I've looked at the specifics of which codes went into (1) or (3).  There are some interactions between the two parts (for example a bad step with Quality -1 gets rejected).

2.  I think it may help the discussion if I could see an example of the batch energy / gradient calculation that you are implementing.  Do you think it would be possible to add a simple unit test that does this?  (For example, two water dimers?)

I was indeed thinking that we would create one thread per OptObject, because the optimization step for each one is computed independently and we could take advantage of parallelism at the level of (at least) a single compute node.  I'm currently not sure how your batch energy / gradient workflow is handling multiple OptObjects and handling optimization steps for each one; adding a test that illustrates how this works would be useful. 

3.  On the future PR: I think your application is a good example of needing a logger to control the flow of printed messages.  Maybe it's a good time to implement some loggers?  I did this once for ForceBalance with the help of a summer student, and we could do it for geomeTRIC too.

**Comment by bgobbi** [comment]
Hi Lee-Ping,
I had to make one more change to the code. I added an ```OPT_STATE``` to the ```OptObject```. The ```step()``` method might actually figure out that it needs to backtrack and skip the energy calculation and evaluation. This is now possible via the state information. The ```Optimize()``` method got two more lines:
```
    optzer = Optimizer(params, xyzout, xyzout2);
    
    optObj  = OptObject(coords, molecule, IC, engine, params.trust, dirname)
        
    # take a step based on grad in optObj
    optzer.step(optObj)
    
    ### Calculate new Energy and Gradient ###
    if optObj.state == OPT_STATE.NEEDS_EVALUATION:
        optObj.calcEnergyForce()
    
    while optzer.evaluateStep(optObj) is OPT_RESULT.NOT_CONVERGED:
        optzer.step(optObj)
        if optObj.state == OPT_STATE.NEEDS_EVALUATION:
            optObj.calcEnergyForce()
     
    return optObj.progress
```

Do you think this is getting to complicated?
I will answer to your previous comments tomorrow. I wanted to get my code working it is less likely that I did miss something.

**Comment by leeping** [comment]
Hello Alberto,

I don't think the solution is too complicated yet, and it's okay for you to get something working for your application locally while we continue working here on making the contribution sustainable.  

I do think there are connections between evaluateStep() and step() that are exposed only when the optimization is sufficiently "difficult".  Rejecting the step is one such situation; rebuilding the coordinate system because it cannot produce a suitable Cartesian step is another possible situation.  One problem is that all of my difficult tests (that involve >100 atoms and >100 steps) have all been run in TeraChem and that takes several hours.  It may be worth the time for me to design some difficult tests that use MM force fields or very approximate QM theories to speed things up.  

- Lee-Ping

**Comment by bgobbi** [comment]
Hi Lee-Ping,
I am not sure how to share the test case for the batch minimization at this point. It would involve adding the NNP code to the project. Here is my optmization loop for you to see my logic though. ```self._nnpComputer.computeBatch(mol_batch)``` is where the forces are computed for the whole batch.
My typical use case would be with batch size around 500 conformation, but on occasion it might be good to be able to do 5000. I think in theory unix can cope with 5000 threads but it is not really the intended use. There would be 5k threads running during the step and evaluation steps.

```python
    def optimizeMols(self, mol_batch):
        """
            Optimize a batch of Mol instances using geomeTRIC.
           
            Parameter
            ---------
               mol_batch (potentially large) batch of cddlib.rdkit.Mol objects 
        """
        optObjs = self.prepareGTOptObjects(mol_batch)

        # Optimization Loop, while not all have completed optmization
        while len(optObjs) > 0:
            nextOptObjs = []
            nextMolsToBeOptimized = []
            
            # take one step, energy and gradient must have been stored in optObj
            for mol, optObj in zip(mol_batch, optObjs):    
                self.gtOptimizer.step(optObj)
                mol.setCoordinates(optObj.currentCoordinatesA())
            
            # compute energies and forces for new coords of whole batch
            efIterator = self._nnpComputer.computeBatch(mol_batch)
                
            for optObj, (mol,e, grad) in zip(optObjs, efIterator):

                if optObj.state == gto.OPT_STATE.NEEDS_EVALUATION:
                    # update energy and forces
                    optObj.engine.energy_AU = e * KCAL_MOL_A_to_AU
                    optObj.engine.grad_AU = grad * KCAL_MOL_A_to_AU
                    
                    # evaluate step
                    optStatus = self.gtOptimizer.evaluateStep(optObj)
                    if optStatus is not gto.OPT_RESULT.NOT_CONVERGED:
                        mol.setProperty("NNP_Energy_kcal_mol",e)
                        mol.setProperty("NNP_Energy_status",str(optStatus))
                        continue

                # add to list for next cycle                    
                nextMolsToBeOptimized.append(mol)
                nextOptObjs.append(optObj)
            
            if len(nextOptObjs) > 0: 
                # step and evaluation completed, next step for remaining conformations
                mol_batch = nextMolsToBeOptimized
                optObjs = nextOptObjs
            
                self.gtOptimizer.step(optObj)
```

I can look into logging, I have no expirience. Do you have a specific logging package in mind?
Thanks,
Alberto

**Comment by leeping** [comment]
Hello Alberto,

Just a quick follow-up, I think it may be possible to create a "fake" batch computer.  The fake batch computer could actually perform individual calculations on a single conformation, but the interface should look like the NNP computer.  

I'm only requesting a test that demonstrates the batch execution logic, the NNP computer does not need to be directly tested.  We don't need to use the true number of conformations (>100) for the test; 2-5 conformations will do.

@dgasmith may have thoughts on testing as well as a good logging package.  I was going to use loggers from the Python standard library.

- Lee-Ping

**Comment by bgobbi** [comment]
This is implemented now. With different splitting but same functionality.
Thanks!
